### PR TITLE
Generate API docs

### LINKED ISSUE #5528: [CT-902] CLI (Top level API) is automatically documented
**User Story:**
As a user I would like to have accurate and easy to read documentation for how to use `dbt-core` as a library.

**Acceptance Criteria:**
An auto-documentation product (probably Sphinx) is set up and documents the top-level api.  This runs as a GHA.

### PR DESCRIPTION
**Pull request by stu-k**
resolves #5528

### Description

Allows for the auto generation of docs for the high level CLI API. This does not also generate those docs via Github Actions, only enables their generation.

### Checklist

- [X] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me
- [X] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [X] I have run this code in development and it appears to resolve the stated issue
- [ ] This PR includes tests, or tests are not required/relevant for this PR
- [ ] I have [opened an issue to add/update docs](https://github.com/dbt-labs/docs.getdbt.com/issues/new/choose), or docs changes are not required/relevant for this PR
- [ ] I have run `changie new` to [create a changelog entry](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md#adding-a-changelog-entry)

### CODE REVIEW

File: core/dbt/cli/main.py [hunk 1]
```diff
@@ -409,4 +409,5 @@ def test(ctx, **kwargs):
 
 # Support running as a module
 if __name__ == "__main__":
+    breakpoint()
```
Comment (ChenyuLInx) [inline core/dbt/cli/main.py:412]: remove before merge?

File: core/dbt/docs/README.rst [hunk 2]
```diff
@@ -0,0 +1,5 @@
+dbt-core
+=======
+
+**dbt-core** An ELT tool for managing your SQL transformations and data models.
```
Comment (ChenyuLInx) [inline core/dbt/docs/README.rst:4]: Just for understanding, is this file used to generate some header for the generated doc?

File: core/dbt/docs/source/_ext/dbt_click.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+            command_section = format_command(cmd)
+            section.extend(command_section)
+        return [section]
+
+    def _get_commands(self, module: str) -> t.List[click.Command]:
+        click_group = load_module(module, self.error)
+        if type(click_group) is not click.Group:
+            raise self.error('Type "click.Group" not supported in dbt_click extension')
+        cmd_strs = [cmd for cmd in click_group.commands]
+        cmd_strs.sort()
```
Comment (ChenyuLInx) [inline core/dbt/docs/source/_ext/dbt_click.py:100]: This doesn't seem to be used?

**Review by ChenyuLInx — COMMENTED** [review]
Looks good over all! Left some questions.

File: core/dbt/docs/source/_ext/dbt_click.py (hunk 3 above)
Comment (stu-k) [reply core/dbt/docs/source/_ext/dbt_click.py:100]: This mutates `cmd_strs`.

File: core/dbt/docs/README.rst (hunk 2 above)
Comment (stu-k) [reply core/dbt/docs/README.rst:4]: I'm actually not sure if this is necessary, this isn't part of what is generated. I'm attempting to remove it and test if everything still works as intended, but I'm having some issues with mypy stubs. I'll check once I resolve those.

File: core/dbt/docs/source/_ext/dbt_click.py (hunk 3 above)
Comment (ChenyuLInx) [reply core/dbt/docs/source/_ext/dbt_click.py:100]: Oh I mean `cmd_strs` itself is not used in other places based on my search

Comment (stu-k) [reply core/dbt/docs/source/_ext/dbt_click.py:100]: Oh, you're right, like 102 should be `for cmd_str in cmd_strs`. Good catch!

File: core/dbt/docs/README.rst (hunk 2 above)
Comment (stu-k) [reply core/dbt/docs/README.rst]: Removed!

**Review by ChenyuLInx — APPROVED** [review]

**Pull request merged**